preprocessor.h

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#ifndef PREPROCESSOR_H
+#define PREPROCESSOR_H
+
+#include <string>
+#include <vector>
+
+#include "ast.h"
+
+struct DirectiveState
+{
+    Directive command;
+    bool condition;
+
+    DirectiveState(Directive command_, bool condition_)
+    {
+        command = command_;
+        condition = condition_;
+    }
+};
+
+class Preprocessor
+{
+    std::vector<DirectiveState> stack_;
+    const std::vector<std::string> defines_;
+
+  public:
+    Preprocessor(const std::vector<std::string>& defines) : defines_(defines)
+    {
+        if (!stack_.empty())
```

Comment:
Is this needed? When constructed, stack will be empty.

Reply:
remove this check